qcportal/qcportal/external/scaffold.py

```diff
@@ -0,0 +1,94 @@
+**before** submission. This means that records are not present.
+
+Other functions are included for similar creation and importation from HDF5 files, although fewer fields
+are retained and transferred in this case.
+
+Organization: Open Molecular Software Foundation: Open Force Field
+Author: Jennifer A Clark
+Date: March 13 2025
+"""
+
+import json
+import bz2
+
+from ..serialization import encode_to_json
+
+
+def to_json(ds, filename="scaffold.json", indent=4, compress=False):
+    """Export a QCFractal dataset to a json file.
+
+    Can be imported with :func:`fom_json` to make a new dataset.
+
+    Args:
+        ds (qcportal.*Dataset): QCFractal dataset
+        filename (str, optional): Filename/path to store output json file. Defaults to "scaffold.json".
+        indent (int, optional): Level of indent for the output json file. Defaults to 4.
+        compress (bool, optional): If True, will compress to bz2. Defaults to False.
+    """
+
+    inputs = [
+        "dataset_type",
+        "name",
+        "description",
+        "tagline",
+        "tags",
+        "group",
+        "provenance",
+        "visibility",
+        "default_tag",
+        "default_priority",
+        "metadata",
+        "owner_group",
+    ]  # Inputs for client.add_dataset(
+    metadata = {key: value for key, value in ds.dict().items() if key in inputs}
+    d = {
+        "metadata": metadata,
+        "entries": {entry.name: entry for entry in ds.iterate_entries()},
+        "specifications": ds.specifications,
+    }
+    d_serializable = encode_to_json(d)
+    
+    if compress:
+        with bz2.open(filename+".bz2", "wt", encoding="utf-8") as f:
+            json.dump(d_serializable, f, ensure_ascii=False, indent=indent)
+    else:
+        with open(filename, "w") as f:
+            json.dump(d_serializable, f, indent=indent)
+
+
+def from_json(filename, client):
+    """Create a QCFractal dataset from a json file.
+
+    Created from output of :func:`to_json`. This allows a user to save the "state"
+    of a dataset before submission.
+
+    Args:
+        filename (str): Filename/path to imported json file.
+        client (qcportal.client.PortalClient): Client to which the dataset will be added.
+
+    Returns:
+        qcportal.*Dataset: QCFractal dataset. This dataset is not submitted in this function.
+    """
+
+    extension = filename.split(".")[-2:]
+    if extension[-1] == "json":
+        with open(filename, "r") as f:
+            ds_dict = json.load(f)
+    elif extension[-1] == "bz2" and extension[-2] == "json":
+        with bz2.open(filename, "rt", encoding="utf-8") as f:
+            ds_dict = json.load(f)
+    else:
+        raise ValueError(f"File extension must be json or json.bz2, not {extension[-2]}.{extension[-1]}")
+
+    ds = client.add_dataset(**ds_dict["metadata"])
+
+    for _, entry in ds_dict["entries"].items():
+        ds.add_entry(**entry)
```

Comment:
Kind of surprised this works. I guess it's just from keeping names consistent :)

It would be safer (and faster?) to build entry object(s) and use ds.add_entries(obj_list).

Every dataset has it's own entry type, but you can get it with `ds._entry_type`. So this might work (going from memory)

```
entry_type = ds._entry_type
....
new_entry = entry_type(**entry)
```

I think specifications will be ok to do your current way, though. Those are consistent across dataset types.

Reply:
The `add_entry` seems to wrap each entry type appropriately. You're right, it cuts the time in half to use.